File: app/org/biobank/domain/participants/Specimen.scala
package org.biobank.domain.participants

import java.time.OffsetDateTime
import java.time.format.DateTimeFormatter
import org.biobank.ValidationKey
import org.biobank.dto.SpecimenDto
import org.biobank.domain._
import org.biobank.domain.containers.{ContainerId, ContainerSchemaPositionId}
import org.biobank.domain.studies.{CollectionSpecimenDefinition, SpecimenDefinitionId, StudyValidations}
import org.biobank.domain.{ConcurrencySafeEntity, DomainValidation}
import org.biobank.infrastructure.EnumUtils._
import org.biobank.services.centres.CentreLocationInfo
import play.api.libs.json._
import scalaz.Scalaz._

/**
 * Represents something that was obtained from a [[domain.participants.Participant Participant]] in a
 * [[domain.studies.Study Study]].
 *
 * A Specimen collected from a [[domain.participants.Participant Participant]] can be created with
 * this aggregate and then added to a [[domain.participants.CollectionEvent CollectionEvent]]. When a
 * specimen is created it must be assigned the corresponding [[domain.studies.SpecimenDefinition
 * SpecimenDefinition]] defined in either the [[domain.participants.CollectionEvent
 * CollectionEvent]] or the specimen link type to which it corresponds .
 */
sealed trait Specimen
    extends ConcurrencySafeEntity[SpecimenId]
    with HasSlug {

  val state: EntityState

  /** The inventory ID assigned to this specimen. */
  val inventoryId: String

  /** The [[domain.studies.CollectionSpecimenDefinition CollectionSpecimenDefinition]] this specimen
    * belongs to, defined by the study it belongs to. */
  val specimenDefinitionId: SpecimenDefinitionId

  /** The [[domain.centres.Centre Centre]] where this specimen was created. */
  val originLocationId: LocationId

  /** The [[domain.centres.Centre Centre]] where this specimen is currently located. */
  val locationId: LocationId

  /** The [[domain.containers.Container Container]] this specimen is stored in. */
  val containerId: Option[ContainerId]

  /**
   * The [[domain.containers.ContainerSchemaPosition ContainerSchemaPosition]] (i.e. position or label) this
   * specimen has in its container.
   */
  val positionId: Option[ContainerSchemaPositionId]

  /**
   * The date and time when the specimen was physically created.
   *
   * Not necessarily when this specimen was added to the application.
   */
  val timeCreated: OffsetDateTime

  /**
   * The amount, in units specified in the [[domain.studies.SpecimenDefinition SpecimenDefinition]], for this
   * specimen.
   */
  val amount: scala.math.BigDecimal

  def createDto(collectionEvent:    CollectionEvent,
                eventTypeName:      String,
                specimenDesc:       CollectionSpecimenDefinition,
                originLocationInfo: CentreLocationInfo,
                locationInfo:       CentreLocationInfo): SpecimenDto =
    SpecimenDto(id                      = this.id.id,
                version                 = this.version,
                timeAdded               = this.timeAdded.format(DateTimeFormatter.ISO_OFFSET_DATE_TIME),
                timeModified            = this.timeModified.map(_.format(DateTimeFormatter.ISO_OFFSET_DATE_TIME)),
                state                   = this.state.id,
                slug                    = this.slug,
                inventoryId             = this.inventoryId,
                collectionEventId       = collectionEvent.id.id,
                specimenDefinitionId    = this.specimenDefinitionId.id,
                specimenDefinitionName  = specimenDesc.name,
                specimenDefinitionUnits = specimenDesc.units,
                originLocationInfo      = originLocationInfo,
                locationInfo            = locationInfo,
                containerId             = this.containerId.map(_.id),
                positionId              = this.positionId.map(_.id),
                timeCreated             = this.timeCreated.format(DateTimeFormatter.ISO_OFFSET_DATE_TIME),
                amount                  = this.amount,
                units                   = specimenDesc.units,
                isDefaultAmount         = (this.amount == specimenDesc.amount),
                eventTypeName           = eventTypeName)

  override def toString: String =
    s"""|${this.getClass.getSimpleName}: {
        |  id:                   $id
        |  version:              $version
        |  timeAdded:            $timeAdded
        |  timeModified:         $timeModified
        |  slug:                 $slug,
        |  inventoryId:          $inventoryId
        |  specimenDefinitionId: $specimenDefinitionId
        |  originLocationId:     $originLocationId
        |  locationId:           $locationId
        |  containerId:          $containerId
        |  positionId:           $positionId
        |  timeCreated:          $timeCreated
        |  amount:               $amount
        |}""".stripMargin
}

object Specimen {
  val usableState: EntityState = new EntityState("usable")
  val unusableState: EntityState = new EntityState("unusable")

  implicit val specimenWrites: Format[Specimen] = new Format[Specimen] {

      override def writes(specimen: Specimen): JsValue =
        ConcurrencySafeEntity.toJson(specimen) ++
        Json.obj(
          "state"                 -> specimen.state.id,
          "slug"                  -> specimen.slug,
          "inventoryId"           -> specimen.inventoryId,
          "specimenDefinitionId" -> specimen.specimenDefinitionId,
          "originLocationId"      -> specimen.originLocationId.id,
          "locationId"            -> specimen.locationId.id,
          "containerId"           -> specimen.containerId,
          "positionId"            -> specimen.positionId,
          "version"               -> specimen.version,
          "timeCreated"           -> specimen.timeCreated,
          "amount"                -> specimen.amount
        )

      override def reads(json: JsValue): JsResult[Specimen] = (json \ "state") match {
          case JsDefined(JsString(usableState.id)) => json.validate[UsableSpecimen]
          case JsDefined(JsString(unusableState.id))  => json.validate[UnusableSpecimen]
          case _ => JsError("error")
        }
    }

  implicit val usableSpecimenReads: Reads[UsableSpecimen] = Json.reads[UsableSpecimen]
  implicit val unusableSpecimenReads: Reads[UnusableSpecimen]   = Json.reads[UnusableSpecimen]

  val sort2Compare: Map[String, (Specimen, Specimen) => Boolean] =
    Map[String, (Specimen, Specimen) => Boolean](
      "inventoryId" -> compareByInventoryId,
      "timeCreated" -> compareByTimeCreated,
      "state"       -> compareByState)

  def compareById(a: Specimen, b: Specimen): Boolean =
    (a.id.id compareTo b.id.id) < 0

  def compareByInventoryId(a: Specimen, b: Specimen): Boolean =
    (a.inventoryId compareTo b.inventoryId) < 0

  def compareByTimeCreated(a: Specimen, b: Specimen): Boolean =
    (a.timeCreated compareTo b.timeCreated) < 0

  def compareByState(a: Specimen, b: Specimen): Boolean =
    (a.state.toString compareTo b.state.toString) < 0
}

trait SpecimenValidations {

  case object InventoryIdInvalid extends ValidationKey

  case object SpecimenDefinitionIdInvalid extends ValidationKey

}

/**
 * A usable specimen is a specimen that can be used for processing.
 */
final case class UsableSpecimen(id:                   SpecimenId,
                                version:              Long,
                                timeAdded:            OffsetDateTime,
                                timeModified:         Option[OffsetDateTime],
                                slug:                 Slug,
                                inventoryId:          String,
                                specimenDefinitionId: SpecimenDefinitionId,
                                originLocationId:     LocationId,
                                locationId:           LocationId,
                                containerId:          Option[ContainerId],
                                positionId:           Option[ContainerSchemaPositionId],
                                timeCreated:          OffsetDateTime,
                                amount:               BigDecimal)
    extends { val state: EntityState = Specimen.usableState }
    with HasSlug
    with Specimen
    with SpecimenValidations
    with ParticipantValidations
    with StudyValidations {

  import org.biobank.domain.DomainValidations._
  import org.biobank.CommonValidations._

  def withInventoryId(inventoryId: String): DomainValidation[Specimen] = {
    validateNonEmptyString(inventoryId, InventoryIdInvalid).map { s =>
      copy(slug         = Slug(inventoryId),
           inventoryId  = inventoryId,
           version      = version + 1,
           timeModified = Some(OffsetDateTime.now))
    }
  }

  def withAmount(amount: BigDecimal): DomainValidation[Specimen] = {
    validatePositiveNumber(amount, AmountInvalid).map { s =>
      copy(amount       = amount,
           version      = version + 1,
           timeModified = Some(OffsetDateTime.now))
    }
  }

  /**
   * Location should belong to a centre.
   */
  def withOriginLocation(id: LocationId): DomainValidation[Specimen] = {
    validateId(id, LocationIdInvalid).map { s =>
      copy(originLocationId = id,
           version          = version + 1,
           timeModified     = Some(OffsetDateTime.now))
    }
  }

  /**
   * Location should belong to a centre.
   */
  def withLocation(id: LocationId): DomainValidation[Specimen] = {
    validateId(id, LocationIdInvalid).map { s =>
      copy(locationId   = id,
           version      = version + 1,
           timeModified = Some(OffsetDateTime.now))
    }
  }

  def withPosition(positionId: ContainerSchemaPositionId): DomainValidation[Specimen] = {
    validateId(positionId, PositionInvalid).map { s =>
      copy(positionId   = Some(positionId),
           version      = version + 1,
           timeModified = Some(OffsetDateTime.now))
    }
  }

  def makeUnusable(): DomainValidation[UnusableSpecimen] = {
    UnusableSpecimen(id                    = this.id,
                     version               = this.version + 1,
                     timeAdded             = this.timeAdded,
                     timeModified          = Some(OffsetDateTime.now),
                     slug                  = this.slug,
                     inventoryId           = this.inventoryId,
                     specimenDefinitionId = this.specimenDefinitionId,
                     originLocationId      = this.originLocationId,
                     locationId            = this.locationId,
                     containerId           = this.containerId,
                     positionId            = this.positionId,
                     timeCreated           = this.timeCreated,
                     amount                = this.amount).successNel[String]
  }
}

object UsableSpecimen
    extends SpecimenValidations
    with ParticipantValidations
    with StudyValidations {

  import org.biobank.CommonValidations._
  import org.biobank.domain.DomainValidations._

  def create(id:                    SpecimenId,
             inventoryId:           String,
             specimenDefinitionId: SpecimenDefinitionId,
             version:               Long,
             originLocationId:      LocationId,
             locationId:            LocationId,
             containerId:           Option[ContainerId],
             positionId:            Option[ContainerSchemaPositionId],
             timeAdded:             OffsetDateTime,
             timeCreated:           OffsetDateTime,
             amount:                BigDecimal)
      : DomainValidation[UsableSpecimen] = {
    validate(id,
             inventoryId,
             specimenDefinitionId,
             version,
             originLocationId,
             locationId,
             containerId,
             positionId,
             amount)
      .map(_ => UsableSpecimen(id                    = id,
                               version               = version,
                               timeAdded             = timeAdded,
                               timeModified          = None,
                               slug                  = Slug(inventoryId),
                               inventoryId           = inventoryId,
                               specimenDefinitionId = specimenDefinitionId,
                               originLocationId      = originLocationId,
                               locationId            = locationId,
                               containerId           = containerId,
                               positionId            = positionId,
                               timeCreated           = timeCreated,
                               amount                = amount))
  }

  def validate(id:                    SpecimenId,
               inventoryId:           String,
               specimenDefinitionId: SpecimenDefinitionId,
               version:               Long,
               originLocationId:      LocationId,
               locationId:            LocationId,
               containerId:           Option[ContainerId],
               positionId:            Option[ContainerSchemaPositionId],
               amount:                BigDecimal)
      : DomainValidation[Boolean] = {
    (validateId(id) |@|
       validateNonEmptyString(inventoryId, InventoryIdInvalid) |@|
       validateId(specimenDefinitionId, SpecimenDefinitionIdInvalid) |@|
       validateVersion(version) |@|
       validateNonEmptyString(originLocationId.id, OriginLocationIdInvalid) |@|
       validateNonEmptyString(locationId.id, LocationIdInvalid) |@|
       validateIdOption(containerId, ContainerIdInvalid) |@|
       validateIdOption(positionId, PositionInvalid) |@|
       validatePositiveNumber(amount, AmountInvalid)) {
      case _ => true
    }
  }

}

/**
 * An Unusable specimen is a specimen that can no longer be used for processing.
 *
 * It may be that the total amount of the spcimen has already been used in processing.
 */
final case class UnusableSpecimen(id:                    SpecimenId,
                                  version:               Long,
                                  timeAdded:             OffsetDateTime,
                                  timeModified:          Option[OffsetDateTime],
                                  slug: Slug,
                                  inventoryId:           String,
                                  specimenDefinitionId: SpecimenDefinitionId,
                                  originLocationId:      LocationId,
                                  locationId:            LocationId,
                                  containerId:           Option[ContainerId],
                                  positionId:            Option[ContainerSchemaPositionId],
                                  timeCreated:           OffsetDateTime,
                                  amount:                BigDecimal)
    extends { val state: EntityState = Specimen.unusableState }
    with HasSlug
    with Specimen {

  def makeUsable(): DomainValidation[UsableSpecimen] = {
    UsableSpecimen(id                    = this.id,
                   version               = this.version + 1,
                   timeAdded             = this.timeAdded,
                   timeModified          = Some(OffsetDateTime.now),
                   slug                  = this.slug,
                   inventoryId           = this.inventoryId,
                   specimenDefinitionId = this.specimenDefinitionId,
                   originLocationId      = this.originLocationId,
                   locationId            = this.locationId,
                   containerId           = this.containerId,
                   positionId            = this.positionId,
                   timeCreated           = this.timeCreated,
                   amount                = this.amount).successNel[String]
  }
}

File: app/assets/javascripts/domain/participants/Specimen.js
/**
 * @author Nelson Loyola <loyola@ualberta.ca>
 * @copyright 2015 Canadian BioSample Repository (CBSR)
 */
define(['lodash', 'tv4', 'sprintf-js'], function(_, tv4, sprintf) {
  'use strict';

  SpecimenFactory.$inject = [
    '$q',
    '$log',
    'ConcurrencySafeEntity',
    'DomainError',
    'biobankApi',
    'centreLocationInfoSchema'
  ];

  /**
   * Factory for Specimens.
   *
   * @param {$q} $q - AngularJS service for asynchronous functions.
   *
   * @param {ConcurrencySafeEntity} ConcurrencySafeEntity - Base class for domain objects.
   *
   * @param {biobankApi} biobankApi - service that communicates with the Biobank server.
   *
   * @returns {Factory} The AngularJS factory function.
   */
  function SpecimenFactory($q,
                           $log,
                           ConcurrencySafeEntity,
                           DomainError,
                           biobankApi,
                           centreLocationInfoSchema) {

    /**
     * Use this contructor to create new Specimens to be persited on the server. Use [create()]{@link
     * domain.participants.Specimen.create} or [asyncCreate()]{@link domain.participants.Specimen.asyncCreate}
     * to create objects returned by the server.
     *
     * @classdesc Represents something that was obtained from a {@link domain.participant.Participant} in a
     * {@link domain.study.Study}.
     *
     * @class
     * @memberOf domain.participants
     * @extends domain.ConcurrencySafeEntity
     *
     * @param {object} [obj={}] - An initialization object whose properties are the same as the members from
     *        this class. Objects of this type are usually returned by the server's REST API.
     *
     * @param {domain.studies.CollectionSpecimenSpec} [specimenSpec] - The specimen spec from the collection
     *        event type this specimen represents. An Undefined value is also valid.
     *
     */
    function Specimen(obj, specimenSpec) {

      /**
       * The unique inventory ID for this specimen.
       *
       * @name domain.participants.Specimen#inventoryId
       * @type {string}
       * @default null
       */
      this.inventoryId = null;

      /**
       * The ID corresponding to the the {@link domain.studies.CollectionSpecimenSpec SpecimenSpec}
       * for this specimen.
       *
       * @name domain.participants.Specimen#specimenSpecId
       * @type {string}
       * @default null
       */
      this.specimenSpecId = null;

      /**
       * The location of the {@link domain.participants.Centre} where this specimen was created.
       * @name domain.participants.Specimen#originLocationId
       * @type {string}
       */
      this.originLocationInfo = null;

      /**
       * The location of the {@link domain.participants.Centre} where this specimen is currently located.
       * @name domain.participants.Specimen#locationId
       * @type {string}
       */
      this.locationInfo = null;

      /**
       * The ID of the {@link domain.participants.Container} this specimen is stored in.
       * @name domain.participants.Specimen#containerId
       * @type {string}
       */

      /**
       * The {@link domain.centres.ContainerSchemaPosition} (i.e. position or label) this specimen has in its
       * container.
       * @name domain.participants.Specimen#positionId
       * @type {string}
       */

      /**
       * The date and time when the specimen was physically created. Not necessarily when this specimen was
       * added to the application.
       * @name domain.participants.Specimen#timeCreated
       * @type {Date}
       */
      this.timeCreated = null;

      /**
       * The amount this specimen contains, in units specified in {@link
       * domain.studies.CollectionSpecimenSpec#units CollectionSpecimenSpec#units}.
       *
       * @name domain.participants.Specimen#amount
       * @type {number}
       */
      this.amount = null;

      /**
       * The state for this specimen. One of: Usable or Unusable.
       * @name domain.participants.Specimen#state.
       * @type {string}
       */
      this.state = null;

      obj = obj || {};
      ConcurrencySafeEntity.call(this, obj);
      _.extend(this, obj);

      if (specimenSpec) {
        this.setSpecimenSpec(specimenSpec);
      }
    }

    Specimen.prototype = Object.create(ConcurrencySafeEntity.prototype);
    Specimen.prototype.constructor = Specimen;

    /**
     * Used for validation.
     */
    Specimen.schema = {
      'id': 'Specimen',
      'type': 'object',
      'properties': {
        'id':               { 'type': 'string' },
        'inventoryId':      { 'type': 'string' },
        'specimenSpecId':   { 'type': 'string' },
        'version':          { 'type': 'integer', 'minimum': 0 },
        'timeAdded':        { 'type': 'string' },
        'timeModified':     { 'type': [ 'string', 'null' ] },
        'originLocationInfo': {
          'type': 'object',
          'items': { '$ref': 'CentreLocationInfo' }
        },
        'locationInfo': {
          'type': 'object',
          'items': { '$ref': 'CentreLocationInfo' }
        },
        'containerId':      { 'type': [ 'string', 'null' ] },
        'postitionId':      { 'type': [ 'string', 'null' ] },
        'timeCreated':      { 'type': 'string' },
        'amount':           { 'type': 'number' },
        'state':            { 'type': 'string' }
      },
      'required': [
        'id',
        'inventoryId',
        'specimenSpecId',
        'version',
        'timeCreated',
        'state',
        'originLocationInfo',
        'locationInfo'
      ]
    };

    /**
     * @private
     */
    Specimen.isValid = function(obj) {
      tv4.addSchema(centreLocationInfoSchema);
      tv4.addSchema(Specimen.schema);
      return tv4.validate(obj, Specimen.schema);
    };

    /**
     * Creates a Specimen, but first it validates #obj to ensure that it has a valid schema.
     *
     * @param {object} [obj={}] - An initialization object whose properties are the same as the members from
     * this class. Objects of this type are usually returned by the server's REST API.
     *
     * @param {CollectionSpecimenSpec} [specimenSpec] - The specimen spec from the collection event type this
     * specimen represents.
     *
     * @returns {Specimen} A new specimen.
     *
     * @see [asyncCreate()]{@link domain.participants.Specimen.asyncCreate} when you need to create
     * a specimen within asynchronous code.
     */
    Specimen.create = function (obj, specimenSpec) {
      if (!Specimen.isValid(obj)) {
        $log.error('invalid object from server: ' + tv4.error);
        throw new DomainError('invalid object from server: ' + tv4.error);
      }

      return new Specimen(obj, specimenSpec);
    };

    /**
     * Creates a Specimen from a server reply but first validates that it has a valid schema.
     *
     * <p>Meant to be called from within promise code.</p>
     *
     * @param {object} [obj={}] - An initialization object whose properties are the same as the members from
     * this class. Objects of this type are usually returned by the server's REST API.
     *
     * @returns {Promise} A new specimen wrapped in a promise.
     *
     * @see [create()]{@link domain.participants.Specimen.create} when not creating a Specimen within
     * asynchronous code.
     */
    Specimen.asyncCreate = function (obj) {
      if (!Specimen.isValid(obj)) {
        $log.error('invalid object from server: ' + tv4.error);
        return $q.reject('invalid object from server: ' + tv4.error);
      }

      return  $q.when(new Specimen(obj));
    };

    /**
     * Retrieves a Specimen from the server.
     *
     * @param {string} id the ID of the specimen to retrieve.
     *
     * @returns {Promise} The specimen within a promise.
     */
    Specimen.get = function (id) {
      if (!id) {
        throw new DomainError('specimen id not specified');
      }

      return biobankApi.get(uri(id)).then(function (reply) {
        return Specimen.asyncCreate(reply);
      });
    };

    /**
     * Retrieves a Specimen with the given inventory ID from the server.
     *
     * @param {string} inventoryId the inventory ID for the specimen.
     *
     * @returns {Promise} The specimen within a promise.
     */
    Specimen.getByInventoryId = function (inventoryId) {
      if (!inventoryId) {
        throw new DomainError('specimen inventory id not specified');
      }

      return biobankApi.get(uri() + '/invid/' + inventoryId).then(function (reply) {
        return Specimen.asyncCreate(reply);
      });
    };

    /**
     * Used to list the specimens contained in a [CollectionEvent]{@link domain.CollectionEvent}.
     *
     * <p>A paged API is used to list specimens. See below for more details.</p>
     *
     * @param {string} ceventId - the CollectionEvent ID that all the specimens belong to.
     *
     * @param {object} options - The options to use to list specimens.
     *
     * @param {string} options.sort - The field to sort the specimens by: one of: <code>id</code>,
     *        <code>timeCreated</code>, <code>state</code>. Use a minus sign prefix to sort in descending
     *        order.
     *
     * @param {number} options.page - The page to retrieve when there are more than <code>options.limit</code>
     *        specimens.
     *
     * @param {number} options.limit - The maximum number of specimens returned per page.
     *
     * @returns {Promise} A promise of {@link biobank.domain.PagedResult} with items of type [Specimen]{@link
     *          domain.Specimen}.
     */
    Specimen.list = function (ceventId, options) {
      var params,
          validKeys = [
            'sort',
            'page',
            'limit'
          ];

      params = _.omitBy(_.pick(options, validKeys), function (value) {
        return value === '';
      });
      return biobankApi.get(uri(ceventId), params).then(function(reply) {
        // reply is a paged result
        var deferred = $q.defer();
        try {
          reply.items = _.map(reply.items, function(obj){
            return Specimen.create(obj);
          });
          deferred.resolve(reply);
        } catch (e) {
          deferred.reject('invalid specimens from server');
        }
        return deferred.promise;
      });
    };

    /**
     * Adds specimens to a collection event.
     *
     * <p>This method should only be used to add collected specimens to a collection event.</p>
     *
     * @param {string} ceventId - the CollectionEvent ID that the specimens will be added to.
     *
     * @param {Array.<Specimens>} - The specimens to add. Only new specimens can be added.
     *
     * @returns {undefined}
     */
    Specimen.add = function (ceventId, specimens) {
      var json = { collectionEventId: ceventId };

      json.specimenData = _.map(specimens, function (specimen) {
        return _.pick(specimen, 'inventoryId', 'specimenSpecId', 'timeCreated', 'locationId', 'amount');
      });
      return biobankApi.post(uri(ceventId), json);
    };

    /**
     * Sets the specimen spec ID for this Specimen.
     *
     * @param {domain.studies.CollectionSpecimenSpec} specimenSpec The specimen specifications associated with
     * this specimen.
     *
     * @returns {undefined}
     */
    Specimen.prototype.setSpecimenSpec = function (specimenSpec) {
      this.specimenSpecId = specimenSpec.uniqueId;
      this.specimenSpec = specimenSpec;
    };

    /**
     * Returns the name for a specimen of this type.
     *
     * @returns {number} The amount used for this specimen.
     */
    Specimen.prototype.name = function () {
      if (_.isUndefined(this.specimenSpec)) {
        throw new DomainError('specimen spec not assigned');
      }
      return this.specimenSpec.name;
    };

    /**
     * Returns the default amount that should be collected for a specimen of this type.
     *
     * @returns {number} The amount used for this specimen.
     */
    Specimen.prototype.defaultAmount = function () {
      if (_.isUndefined(this.specimenSpec)) {
        throw new DomainError('specimen spec not assigned');
      }
      return this.specimenSpec.amount;
    };

    /**
     * Whether or not the amount for this specimen is the default amount.
     *
     * @returns {boolean} True if the amount is the default.
     */
    Specimen.prototype.isDefaultAmount = function () {
      if (_.isUndefined(this.specimenSpec)) {
        throw new DomainError('specimen spec not assigned');
      }
      return (this.amount === this.specimenSpec.amount);
    };

    /**
     * Returns the units a specimen of this type.
     *
     * @returns {string} The units used for this specimen.
     */
    Specimen.prototype.units = function () {
      if (_.isUndefined(this.specimenSpec)) {
        throw new DomainError('specimen spec not assigned');
      }
      return this.specimenSpec.units;
    };

    /**
     * Removes the specimen from the system.
     *
     * @return {Promise} Resolves to true if the specimen was removed successfully.
     */
    Specimen.prototype.remove = function () {
      var url = sprintf.sprintf('%s/%s/%d', uri(this.collectionEventId), this.id, this.version);
      return biobankApi.del(url);
    };

    function uri(/* specimenId */) {
      var specimenId,
          result = '/participants/cevents/spcs',
          args = _.toArray(arguments);

      if (args.length > 0) {
        specimenId = args.shift();
        result += '/' + specimenId;
      }

      return result;
    }

    /** return constructor function */
    return Specimen;
  }

  return SpecimenFactory;
});
